### PR TITLE
Distinguish events and delegates 

### LINKED ISSUE #245: Events: Distinguish Events from Delegates
This is a sub-task for #64 

This article will discuss how to distinguish between delegates and events in your designs.

### PR DESCRIPTION
**Pull request by BillWagner**
This fixes #245.

This must be merged after PR #246.

### CODE REVIEW

File: docs/languages/csharp/distinguish-delegates-events.md [hunk 1]
```diff
@@ -1,8 +1,80 @@
...
+method.
+
+Notice that these two heuristics may often both be present: If your
+delegate method returns a value, it will likely impact the algorithm
+in some way.
+
+# Event Listeners Often Have Longer Lifetimes 
+
+This is a slightly weaker justification. However, you may find that
+event based designs are more natural when the event source will be
```
Comment (ghogen) [inline docs/languages/csharp/distinguish-delegates-events.md]: event-based

File: docs/languages/csharp/distinguish-delegates-events.md [hunk 2]
```diff
@@ -1,8 +1,80 @@
...
+raising events over a long period of time. You can see examples of
+this for UX controls on many systems. Once you subscribe to an event,
+the event source may raise events throughout the lifetime of the program.
+(You can unsubscribe from events when you no longer need them.)
+
+Contrast that with many delegate based designs, where a delegate is
```
Comment (ghogen) [inline docs/languages/csharp/distinguish-delegates-events.md]: delegate-based

**Comment by BillWagner** [comment]
responded to comments, rebased, and squashed.

File: docs/languages/csharp/distinguish-delegates-events.md [hunk 3]
```diff
@@ -1,8 +1,80 @@
...
+support. They both support similar syntax for adding and removing
+handlers. Finally, raising an event and calling a delegate use exactly the same method call syntax. They even both support the same `Invoke()`
+method syntax for use with the `?.` operator.
+
+With all those similarities, it is easy to have trouble determining when
+to use which.
+
+# Listening to Events is Optional
+
+The most important consideration in determining which language feature
```
Comment (cartermp) [inline docs/languages/csharp/distinguish-delegates-events.md]: I'd consider rewording this paragraph to flip the use of negatives.  "If your code requires an attached subscriber to complete its task [...]" and so on.  I had some trouble parsing the double negatives.

File: docs/languages/csharp/distinguish-delegates-events.md [hunk 4]
```diff
@@ -1,8 +1,80 @@
...
+to use is whether or not there must be an attached subscriber. If your
+code can't complete its tasks without an attached subscriber, you should
+use a design based on delegates. If the absence of any subscribers
+does not affect your code's ability to complete, you should use a
+design based on events. 
+
+Consider the examples built during this section. The code you built
```
Comment (cartermp) [inline docs/languages/csharp/distinguish-delegates-events.md]: The double negatives in this paragraph may also be worth flipping to a use of positives.

File: docs/languages/csharp/distinguish-delegates-events.md [hunk 5]
```diff
@@ -1,8 +1,80 @@
...
+listening to the events. They both use designs based on events.
+
+# Return Values Require Delegates
+
+Another consideration is the method prototype you would want for your
+delegate method. As you've seen, the delegates used for events all
+have a void return type. You've also seen that there are idioms to
+create event handlers that do pass information back to event sources
+through modifying properties of the event argument object. While these
+idioms do work, they are not as natural as return a value from a
```
Comment (cartermp) [inline docs/languages/csharp/distinguish-delegates-events.md]: returning

**Comment by cartermp** [comment]
LGTM modulo feedback.

**Comment by cartermp** [comment]
@BillWanger feel free to cast SquashAndMerge with your newfound powers!

**Pull request merged**